storey/flow.py

```diff
@@ -608,63 +611,98 @@ async def _handle_completed(self, event, response):
+            if callable(key):
+                self.key_extractor = key
+            elif isinstance(key, str):
+                self.key_extractor = lambda element: element[key]
 
     async def _emit(self, batch, batch_time):
         raise NotImplementedError
 
     async def _terminate(self):
         pass
 
-    async def _sleep_and_emit(self):
-        await asyncio.sleep(self._timeout_secs)
-        await self._emit_batch()
+    async def _sleep_and_emit(self, key: Optional[str] = None):
+        time_delta = (
+            datetime.now(timezone.utc) - self._batch_time[key]
+        ).total_seconds()
+        if time_delta < self._timeout_secs:
+            await asyncio.sleep(self._timeout_secs - time_delta)
+        await self._emit_batch(key)
+        self._timeout_task = None
 
     def _event_to_batch_entry(self, event):
         return self._get_event_or_body(event)
 
     async def _do(self, event):
         if event is _termination_obj:
-            if self._timeout_task and not self._timeout_task.cancelled():
+            if self._timeout_task is not None and not self._timeout_task.cancelled():
                 self._timeout_task.cancel()
-            await self._emit_batch()
+            await self._emit_all()
             await self._terminate()
             return await self._do_downstream(_termination_obj)
-        else:
-            if len(self._batch) == 0:
-                self._batch_time = event.time
-                if self._timeout_secs:
-                    self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit())
-
-            self._event_count = self._event_count + 1
-            self._batch.append(self._event_to_batch_entry(event))
-
-            if self._event_count == self._max_events:
-                if self._timeout_task and not self._timeout_task.cancelled():
-                    self._timeout_task.cancel()
-                await self._emit_batch()
-
-    async def _emit_batch(self):
-        if len(self._batch) > 0:
-            batch_to_emit = self._batch
-            batch_time = self._batch_time
-            self._batch = []
-            self._batch_time = None
-            self._event_count = 0
 
+        key = None
+        if self._group_events:
+            key = self.key_extractor(event.body) if self.key_extractor else event.key
+
+        if len(self._batch[key]) == 0:
+            self._batch_time[key] = event.time
+
+        if self._timeout_secs and self._timeout_task is None:
+            self._init_timeout_task()
+
+        self._event_count[key] = self._event_count[key] + 1
+        self._batch[key].append(self._event_to_batch_entry(event))
+
+        if self._event_count[key] == self._max_events:
+            if self._timeout_task is not None and not self._timeout_task.cancelled():
+                self._timeout_task.cancel()
+            await self._emit_batch(key)
+
+    def _init_timeout_task(self):
+        for key in self._batch.keys():
```

Comment:
Aren't you creating a task per key here? we decided to only have one task for the "next" key to emit

Reply:
There's a break right after it. could have wrapped `self._batch.keys()` with a list, then take the first element, but that wouldn't be less messy.

Reply:
oh right, why don't just do `key = next(iter(self._batch))`

Reply:
Nice, will do